### PR TITLE
Added option to spade to use a 3-dimensional pattern spectrum to test

### PR DESCRIPTION
**Pull request by pietroquaglio**
Added option to spade to use a 3-dimensional pattern spectrum to test separately patterns with different durations. If `spectrum='3d#'` the signature of a pattern is not anymore (number of spikes, number of occurrences), but (number of spikes, number of occurrences, duration).

### CODE REVIEW

**Comment by alperyeg** [comment]
@stellalessandra can you already start the review process despite the errors. I will also have a look. Thank you.

File: elephant/spade.py [hunk 1]
```diff
@@ -300,79 +334,61 @@ def spade(data, binsize, winlen, min_spikes=2, min_occ=2, min_neu=1,
...
                                                             winlen, binsize,
                                                             pv_spec,
                                                             data[0].t_start)
         else:
             output['patterns'] = concepts
         return output
+    # rank!= returning a string message
     else:
-        return []
+        return 'Results returned only for rank==0'
```
[the remark this answers is not in the document]
Comment (pietroquaglio) [reply elephant/spade.py]: I thought to return a meaningful message to explain why the output is not returned

File: elephant/spade.py [hunk 2]
```diff
@@ -458,11 +486,11 @@ def concepts_mining(data, binsize, winlen, min_spikes=2, min_occ=2,
     # spikes in a window
     if max_spikes is None:
         max_spikes = np.max((int(np.max(np.sum(rel_matrix, axis=1))),
-                            min_spikes + 1))
+                             min_spikes + 1))
     # By default, set maximum number of occurrences to number of non-empty
     # windows
     if max_occ is None:
-        max_occ = int(np.sum(np.sum(rel_matrix, axis=1)>0))
+        max_occ = int(np.sum(np.sum(rel_matrix, axis=1) > 0))
```
[the remark this answers is not in the document]
Comment (pietroquaglio) [reply elephant/spade.py]: The function _fpgrowth() takes an integer value in input

File: elephant/spade.py [hunk 3]
```diff
@@ -667,41 +701,68 @@ def _fpgrowth(transactions, min_c=2, min_z=2, max_z=None,
...
-            if rel_matrix is not None:
-                extent = tuple(np.where(
-                    np.prod(rel_matrix[:, intent], axis=1) == 1)[0])
-            concepts.append((intent, extent))
-            if report == '#':
+            # Removing subset occurring in different window positions
+            keep_concept = True
+            c_idx = 0
+            while keep_concept:
+                intent_comp, supp_comp = fpgrowth_output[c_idx]
```
[the remark this answers is not in the document]
Comment (pietroquaglio) [reply elephant/spade.py]: They are the intent and support of the other concept to which the current concept of the for loop is compared to

**Comment by pietroquaglio** [comment]
All the comments have been addressed in the second last commits, excluding only the one in which I replied directly here explicitely

File: elephant/spade.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alperyeg) [reply elephant/spade.py]: I agree with @stellalessandra. Rather I would give out a warning and return None.

File: elephant/spade.py [hunk 4]
```diff
@@ -492,18 +522,22 @@ def concepts_mining(data, binsize, winlen, min_spikes=2, min_occ=2,
...
     binned spike trains
     Parameters
     ----------
     binary_matrix : numpy.array
-        Binary matrix containing the binned spike trais
+        Binary matrix containing the binned spike trains
     winlen : int
         Length of the binsize used to bin the data
+    only_windows_with_first_spike : bool
+        Whether to consider every window or only the one with a spike in the
```
Comment (alperyeg) [inline elephant/spade.py]: I think some motivation for the parameter is needed, like why one should consider it to be True.

File: elephant/spade.py [hunk 5]
```diff
@@ -774,36 +846,69 @@ def _fast_fca(context, min_c=2, min_z=2, max_z=None,
...
         intent = tuple(fca_concept.intent)
         extent = tuple(fca_concept.extent)
+        supp = len(extent)
+        # Removing subset occurring in different window positions
+        keep_concept = True
+        c_idx = 0
+        while keep_concept:
+            intent_comp = tuple(fca_concepts[c_idx].intent)
+            supp_comp = len(tuple(fca_concepts[c_idx].extent))
+            # print(set(np.array(intent_comp) % winlen).issuperset(set(
```
Comment (alperyeg) [inline elephant/spade.py]: Remove code comment or make it optional.

File: elephant/spade.py [hunk 6]
```diff
@@ -994,8 +1124,35 @@ def _fdr(pvalues, alpha):
...
+
+    Parameters
+    ----------
+    pvalues: list
+       list of p-values, each corresponding to a statistical test
+    alpha: float
+       significance level
+
+    Returns
+    ------
```
Comment (alperyeg) [inline elephant/spade.py]: `-` is missing under Returns.

File: elephant/spade.py [hunk 7]
```diff
@@ -994,8 +1124,35 @@ def _fdr(pvalues, alpha):
...
+    Returns a triplet containing:
+    * an array of bool, indicating for each p-value whether it was
```
Comment (alperyeg) [inline elephant/spade.py]: The triplet is not explained.

File: elephant/spade.py [hunk 8]
```diff
@@ -1041,27 +1209,51 @@ def test_signature_significance(pvalue_spectrum, alpha, corr='', report='#'):
         tests = x_array[:, -1] <= alpha * 1. / len(pvalue_spectrum)
     elif corr in ['f', 'fdr']:  # or with FDR correction
         tests, pval, rank = _fdr(x_array[:, -1], alpha=alpha)
+    elif corr in ['hb', 'holm_bonf']:
+        tests = _holm_bonferroni(x_array[:, -1], alpha=alpha)
     else:
-        raise AttributeError("corr must be either '', 'b'('bonf') or 'f'('fdr')")
-
+        raise AttributeError(
+            "corr must be either '', 'b'('bonf') or 'f'('fdr')")
```
Comment (alperyeg) [inline elephant/spade.py]: Is `corr` a parameter?

File: elephant/spade.py [hunk 9]
```diff
@@ -1106,15 +1298,15 @@ def approximate_stability(concepts, rel_matrix, n_subsets, delta=0, epsilon=0):
         epsilon: absolute error
         Default: 0
 
-    Output
+    Returns
```
Comment (alperyeg) [inline elephant/spade.py]: `-` is missing under Returns

File: elephant/test/test_spade.py [hunk 10]
```diff
@@ -243,4 +433,4 @@ def suite():
 if __name__ == "__main__":
     runner = unittest.TextTestRunner(verbosity=2)
     runner.run(suite())
-    globals()
\ No newline at end of file
+    globals()
```
Comment (alperyeg) [inline elephant/test/test_spade.py]: Do you need the globals()? For what?

File: elephant/spade.py [hunk 11]
```diff
@@ -732,36 +801,39 @@ def _fast_fca(context, min_c=2, min_z=2, max_z=None,
...
          Minimum number of neurons in a sequence to considered a
          potential pattern.
          Default: 1
 
-    Returns:
+    Returns
     --------
-    returns:
-    * If report != '#':
+    * If report == 'a':
```
Comment (alperyeg) [inline elephant/spade.py]: I would change the format here. To something like: 
description 
then the if else under the description

File: elephant/spade.py (hunk 8 above)
Comment (pietroquaglio) [reply elephant/spade.py]: Yes it is, should I make this explicit?

File: elephant/spade.py (hunk 11 above)
Comment (pietroquaglio) [reply elephant/spade.py]: I realized it was just wrong, I updated according to the function concept_mining, which has the same output

File: elephant/test/test_spade.py (hunk 10 above)
Comment (pietroquaglio) [reply elephant/test/test_spade.py]: Removed, it shouldn't have been there

File: elephant/spade.py (hunk 8 above)
Comment (alperyeg) [reply elephant/spade.py]: Yes, just for readability would be nice to have e.g. "Parameter `corr` must be ..."

**Comment by alperyeg** [comment]
@stellalessandra are you finished with the review? I would like to merge if @pietroquaglio is through regarding the latest comments.

**Comment by stellalessandra** [comment]
@alperyeg yes, I am done!

File: elephant/spade.py (hunk 8 above)
Comment (pietroquaglio) [reply elephant/spade.py]: Done!

**Pull request merged**